%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% RPREP: Ribosomal Profiling Reports, an open-source cloud-enabled framework for reproducible
% Ribosomal Profiling data processing, analysis, and result reporting
% 
% https://github.com/emmesgit/RPREP
%
% Copyright (C) 2020 The Emmes Company L.L.C. 
% 
% This program is free software that contains third party software subject to various licenses, 
% namely, the GNU General Public License version 3 (or later), the GNU Affero General Public License 
% version 3 (or later), and the LaTeX Project Public License v.1.3(c). A list of the software contained 
% in this program, including the applicable licenses, can be accessed here: 
% 
% https://github.com/emmesgit/RPREP/blob/master/SOFTWARE.xlsx  
% 
% You can redistribute and/or modify this program, including its components, only under the terms of 
% the applicable license(s).  
%
% To cite this software, please reference <DOI>
%
% This program is distributed in the hope that it will be useful, but "as is," WITHOUT ANY WARRANTY; 
% without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
%
% Program:  rp-00-qc-normalization-figures.Rnw
% Version:  RPREP 1.0.1
% Author:   Travis L. Jensen, Johannes B. Goll, Sami R. Cherikh, and William F. Hooper
% Purpose:  Knitr Code to build QC/normalization figures
% Input:    source/r/00-qc-normalization/qc-boxplots.r
%           source/r/00-qc-normalization/qc-bam-star-plots.r
%           source/r/00-qc-normalization/tmm-box-plots.r
%           source/r/00-qc-normalization/tmm-box-plots.r
%           source/r/00-qc-normalization/tmm-cdf-plots.r
%           source/r/00-qc-normalization/reverse-ecdf-plot.r
% Output:	N/A
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Import analysis variables
<<rp_variables, echo=FALSE, results='hide', message=FALSE>>=
reportAssayFlag = 'ribosomal_profiling'
source('../r/init-analysis.r')
@
%--------------------------------------------------------------------
%
% QC BOXPLOTS
%
%---------------------------------------------------------------------
<<rp_qc_boxplots,fig.height=7.5,fig.cap=paste('Boxplots of human reference genome alignment ',bpLabel,' statistics (RP, All specimen types).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/r/00-qc-normalization/qc-boxplots.r',sep=''));
@

%--------------------------------------------------------------------
%
% BAM QC - STARPLOTS
%
%---------------------------------------------------------------------
<<rp_star_plots,fig.cap=paste('Starplots of human reference genome alignment statistics (RP, ',spcLabls,').',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/r/00-qc-normalization/qc-bam-star-plots.r',sep=''));
@

%--------------------------------------------------------------------
%
% TMM NORMALIZATION QC
%
%---------------------------------------------------------------------
<<rp_tmm_box_plots,fig.cap=paste('Boxplots of \\logtwo counts per million ',rep(c('before','after'),each=length(times.knitr.nrm.lab)),' TMM normalization ',rep(times.knitr.nrm.lab,2),'(RP, All specimen types).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/r/00-qc-normalization/tmm-box-plots.r',sep=''));
@

<<rp_tmm_cdf_plots,fig.width=6,fig.height=7.4,out.width="1.0\\textwidth",fig.cap=paste('Empirical cumulative distribution function plots of \\logtwo counts per million before and after TMM normalization ',times.knitr.nrm.lab,'(RP, All specimen types).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/r/00-qc-normalization/tmm-cdf-plots.r',sep=''));
@

%--------------------------------------------------------------------
%
% Reverse ECDF Plot by gene espression cutoff values
%
%---------------------------------------------------------------------
<<rp_recdf_plot,fig.width=9,fig.height=6,out.width="1.0\\textwidth",fig.cap=paste('Reverse empirical cumulative distribution function plots of ',gene_filter_method,' gene expression levels across study samples (RP, All specimen types). The x-axis represents the \\logtwo count per million cut off for identifying lowly expressed genes. The y-axis shows the percentage of all genes whose ',gene_filter_method,' gene expression level across all study samples exceeds the respective cut off.  The grey box indicates the target range of genes to be selected (between ',flt.gene.range[1],' and ',flt.gene.range[2],' genes). The black vertical lines represent the specified cut off (',flt.lcpm.cut,' \\logtwo counts per million).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/r/00-qc-normalization/reverse-ecdf-plot.r',sep=''));
@

%--------------------------------------------------------------------
%
% Gene body coverage plots
%
%---------------------------------------------------------------------
<<rp_gbc_plot,fig.width=9,fig.height=6,out.width="1.0\\textwidth",fig.cap=paste('Gene body coverage plots (RP).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste(src.dir,'/../../r/gene-coverage-bias-plots.r',sep=''));
@

%--------------------------------------------------------------------
%
% Read length distribution plot
%
%---------------------------------------------------------------------
<<rp_el_plot,fig.width=5,fig.height=6,out.width="0.8\\textwidth",fig.cap=paste('Read length distribution plots (RP).',sep='')>>=
reportAssayFlag = 'ribosomal_profiling'
source(paste0(src.dir,'/../../r/read-length-distribution-figures.r'))
@

