%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% RSEQREP: RNA-Seq Reports, an open-source cloud-enabled framework for reproducible
% RNA-Seq data processing, analysis, and result reporting
% 
% https://github.com/emmesgit/RSEQREP
%
% Copyright (C) 2017 The Emmes Corporation 
% 
% This program is free software that contains third party software subject to various licenses, 
% namely, the GNU General Public License version 3 (or later), the GNU Affero General Public License 
% version 3 (or later), and the LaTeX Project Public License v.1.3(c). A list of the software contained 
% in this program, including the applicable licenses, can be accessed here: 
% 
% https://github.com/emmesgit/RSEQREP/blob/master/SOFTWARE.xlsx  
% 
% You can redistribute and/or modify this program, including its components, only under the terms of 
% the applicable license(s).  
%
% This program is distributed in the hope that it will be useful, but "as is," WITHOUT ANY WARRANTY; 
% without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
%
% To cite this software, please reference doi:10.12688/f1000research.13049.1
%
% Program:  rp-rseq-comparison-figures.Rnw
% Version:  RPREP 1.0.0
% Author:   Travis L. Jensen, Johannes B. Goll, Sami R. Cherikh, and William F. Hooper
% Purpose:  Incorporate translation efficiency tables into the report
% Input:    source/r/plt-rnaseq-vs-rp-scatter-exp.r
%			source/r/plt-rnaseq-vs-rp-scatter-lfc.r
% Output:   N/A
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%--------------------------------------------------------------------
%
% EXPRESSION / LFC SCATTERPLOTS
%
%---------------------------------------------------------------------
<<rp_rseq_exp_scatter,fig.width=9,fig.height=12,dev='CairoPNG',dpi=300,out.width="0.95\\textwidth",fig.cap='Scatterplots of \\logtwo expression between shared RNA-Seq and Ribosomal Profiling samples averaged by specimen type, treatment group, and timepoint. The union of the RNA-Seq and Ribosomal Profiling LCPM filtered gene sets are shown. n = number of samples.'>>=
source(paste0(src.dir,'/../../r/plt-rnaseq-vs-rp-scatter-exp.r'))
@
<<rp_rseq_lfc_scatter,fig.width=9,fig.height=12,dev='CairoPNG',dpi=300,out.width="0.95\\textwidth",fig.cap='Scatterplots of the union of differentially expressed genes \\logtwo fold change between shared RNA-Seq and Ribosomal Profiling samples averaged by specimen type, treatment group, and timepoint. n = number of samples.'>>=
source(paste0(src.dir,'/../../r/plt-rnaseq-vs-rp-scatter-lfc.r'))
@
<<rp_rseq_exp_scatter_ind,fig.width=9,fig.height=12,dev='CairoPNG',dpi=300,out.width="0.95\\textwidth",fig.cap='Scatterplots of \\logtwo expression scatterplots between shared RNA-Seq and Ribosomal Profiling by sample. The union of the RNA-Seq and Ribosomal Profiling LCPM filtered gene sets are shown.'>>=
source(paste0(src.dir,'/../../r/plt-rnaseq-vs-rp-scatter-exp-ind.r'))
@
<<rp_rseq_lfc_scatter_ind,fig.width=9,fig.height=12,dev='CairoPNG',dpi=300,out.width="0.95\\textwidth",fig.cap='Scatterplots of  the union of differentially expressed genes \\logtwo fold change scatterplots between shared RNA-Seq and Ribosomal Profiling  by sample.'>>=
source(paste0(src.dir,'/../../r/plt-rnaseq-vs-rp-scatter-lfc-ind.r'))
@